File: examples/dekita/.claude/hooks/parallel-edit-conflict-check.py
#!/usr/bin/env python3
"""sibling fork-session間での同一ファイル編集を警告する。

Why:
    複数のfork-sessionが同一ファイルを編集すると、マージ時に
    コンフリクトが発生する。事前に警告し調整を促す。

What:
    - 編集対象ファイルを取得
    - sibling fork-sessionの変更ファイル一覧を取得
    - 同一ファイルを編集中のsiblingがあれば警告

Remarks:
    - 警告型フック（ブロックしない、systemMessageで警告）
    - PreToolUse:Edit/Writeで発火（fork-sessionのみ）
    - session-worktree-statusは起動時のみ（リアルタイム警告との違い）
    - パス正規化でworktree間の比較を可能に

Changelog:
    - silenvx/dekita#2513: フック追加
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

# Add hooks directory to path
HOOKS_DIR = Path(__file__).parent
sys.path.insert(0, str(HOOKS_DIR))

from lib.execution import log_hook_execution
from lib.session import is_fork_session, parse_hook_input
from lib.session_graph import get_active_worktree_sessions


def get_target_file(tool_input: dict) -> str | None:
    """Extract target file path from tool input.

    Args:
        tool_input: The tool_input from hook input.

    Returns:
        The file path being edited, or None if not found.
    """
    return tool_input.get("file_path")


def normalize_path(file_path: str) -> str:
    """Normalize file path for comparison.

    Converts absolute paths to relative paths from repo root.
    Handles both worktree paths and main repo paths.

    Args:
        file_path: The file path to normalize.

    Returns:
        Normalized relative path.
    """
    path = Path(file_path)
    parts = path.parts

    # Look for .worktrees pattern (handles worktree paths)
    for i, part in enumerate(parts):
        if part == ".worktrees":
            # Skip .worktrees/<name>/ prefix (i is .worktrees, i+1 is name, i+2+ is path)
            if i + 2 < len(parts):
                return str(Path(*parts[i + 2 :]))
            return str(path)

    # For non-worktree paths, try to find a common repo root indicator
    # Look for common project markers (.git, .claude, src, etc.)
    repo_markers = {".git", ".claude", "src", "frontend", "worker", "shared"}
    for i, part in enumerate(parts):
        if part in repo_markers and i > 0:
            # Return path from this point onwards
            return str(Path(*parts[i:]))

    return str(path)


def find_conflicting_worktrees(
    target_file: str,
    active_sessions: dict,
) -> list[dict]:
    """Find worktrees that have the same file in their changed files.

    Args:
        target_file: The file being edited.
        active_sessions: Dict from get_active_worktree_sessions.

    Returns:
        List of conflicting worktree info dicts.
    """
    conflicts = []
    normalized_target = normalize_path(target_file)

    # Check sibling worktrees for conflicts
    sibling_worktrees = active_sessions.get("sibling", [])
    for info in sibling_worktrees:
        # Check if target file is in this worktree's changed files
        for changed_file in info.changed_files:
            if normalize_path(changed_file) == normalized_target:
                conflicts.append(
                    {
                        "issue_number": info.issue_number,
                        "path": str(info.path),
                        "changed_files": sorted(info.changed_files)[:5],
                        "total_files": len(info.changed_files),
                    }
                )
                break

    return conflicts


def format_warning(target_file: str, conflicts: list[dict]) -> str:
    """Format the conflict warning message.

    Args:
        target_file: The file being edited.
        conflicts: List of conflicting worktree info.

    Returns:
        Formatted warning message.
    """
    lines = ["⚠️ 並行編集の競合可能性:\n"]
    lines.append(f"編集対象: {normalize_path(target_file)}\n")
    lines.append("同一ファイルを編集中のsibling session:")

    for conflict in conflicts:
        issue_str = (
            f"Issue #{conflict['issue_number']}"
            if conflict["issue_number"]
            else conflict["path"].split("/")[-1]
        )
        lines.append(f"  - {issue_str}")
        lines.append(f"    Worktree: {conflict['path']}")

        files = conflict["changed_files"]
        total_files = conflict["total_files"]
        files_str = ", ".join(files[:3])
        if total_files > 3:
            files_str += f" (+{total_files - 3} more)"
        lines.append(f"    変更中: {files_str}")
        lines.append("")

    lines.append("マージ時にコンフリクトが発生する可能性があります。")
    lines.append("Tip: 競合を避けるため、siblingセッションと調整するか、")
    lines.append("     別の独立したIssueに着手することを検討してください。")

    return "\n".join(lines)


def main() -> None:
    """PreToolUse hook for Edit/Write tools.

    Warns when editing a file that sibling fork-sessions are also modifying.
    """
    result = {"decision": "approve"}

    try:
        data = parse_hook_input()
        session_id = data.get("session_id", "")
        source = data.get("source", "")
        transcript_path = data.get("transcript_path")
        tool_input = data.get("tool_input", {})

        # Only run for fork-sessions
        if not is_fork_session(session_id, source, transcript_path):
            log_hook_execution("parallel-edit-conflict-check", "approve", "Not a fork-session")
            print(json.dumps(result))
            return

        # Get target file
        target_file = get_target_file(tool_input)
        if not target_file:
            log_hook_execution("parallel-edit-conflict-check", "approve", "No target file")
            print(json.dumps(result))
            return

        # Get active worktree sessions
        try:
            active_sessions = get_active_worktree_sessions(session_id, transcript_path)
        except Exception:
            # Fail silently - don't block on errors
            log_hook_execution("parallel-edit-conflict-check", "approve", "Error getting sessions")
            print(json.dumps(result))
            return

        # Find conflicts
        conflicts = find_conflicting_worktrees(target_file, active_sessions)

        if conflicts:
            warning = format_warning(target_file, conflicts)
            result["systemMessage"] = warning

            log_hook_execution(
                "parallel-edit-conflict-check",
                "approve",
                f"Warning: {len(conflicts)} conflicting worktree(s)",
                {"target_file": target_file, "conflicts": conflicts},
            )
        else:
            log_hook_execution("parallel-edit-conflict-check", "approve", "No conflicts")

    except Exception as e:
        error_msg = f"Hook error: {e}"
        print(f"[parallel-edit-conflict-check] {error_msg}", file=sys.stderr)
        log_hook_execution("parallel-edit-conflict-check", "approve", error_msg)

    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/parallel_edit_conflict_check.ts
#!/usr/bin/env bun
/**
 * sibling fork-session間での同一ファイル編集を警告する。
 *
 * Why:
 *   複数のfork-sessionが同一ファイルを編集すると、マージ時に
 *   コンフリクトが発生する。事前に警告し調整を促す。
 *
 * What:
 *   - 編集対象ファイルを取得
 *   - sibling fork-sessionの変更ファイル一覧を取得
 *   - 同一ファイルを編集中のsiblingがあれば警告
 *
 * State:
 *   - reads: transcript files
 *   - reads: git worktree status
 *
 * Remarks:
 *   - 警告型フック（ブロックしない、systemMessageで警告）
 *   - PreToolUse:Edit/Writeで発火（fork-sessionのみ）
 *   - session-worktree-statusは起動時のみ（リアルタイム警告との違い）
 *   - パス正規化でworktree間の比較を可能に
 *
 * Changelog:
 *   - silenvx/dekita#2513: フック追加
 *   - silenvx/dekita#3162: TypeScriptに移植
 */

import { basename } from "node:path";
import { formatError } from "../lib/format_error";
import { getOriginDefaultBranch } from "../lib/git";
import { logHookExecution } from "../lib/logging";
import { isForkSession, parseHookInput } from "../lib/session";
import { asyncSpawn } from "../lib/spawn";

const HOOK_NAME = "parallel-edit-conflict-check";

interface WorktreeInfo {
  path: string;
  issueNumber: string | null;
  changedFiles: string[];
}

interface ConflictInfo {
  issueNumber: string | null;
  path: string;
  changedFiles: string[];
  totalFiles: number;
}

/**
 * Extract target file path from tool input.
 */
function getTargetFile(toolInput: Record<string, unknown>): string | null {
  return (toolInput.file_path as string) ?? null;
}

/**
 * Normalize file path for comparison.
 *
 * Converts absolute paths to relative paths from repo root.
 * Handles both worktree paths and main repo paths.
 */
function normalizePath(filePath: string): string {
  // If path is already relative (does not start with /), return as is
  if (!filePath.startsWith("/")) {
    return filePath;
  }

  const parts = filePath.split("/").filter((p) => p);

  // Look for .worktrees pattern (handles worktree paths)
  for (let i = 0; i < parts.length; i++) {
    if (parts[i] === ".worktrees") {
      // Skip .worktrees/<name>/ prefix
      if (i + 2 < parts.length) {
        return parts.slice(i + 2).join("/");
      }
      return filePath;
    }
  }

  // For non-worktree paths, try to find a common repo root indicator
  const repoMarkers = new Set([".git", ".claude", "src", "frontend", "worker", "shared"]);
  for (let i = 0; i < parts.length; i++) {
    if (repoMarkers.has(parts[i]) && i > 0) {
      return parts.slice(i).join("/");
    }
  }

  return filePath;
}

/**
 * Get worktree list with their paths and lock status.
 * Parses lock status in a single pass to avoid N+1 queries.
 */
async function getWorktreeList(): Promise<Array<{ path: string; name: string; locked: boolean }>> {
  try {
    const result = await asyncSpawn("git", ["worktree", "list", "--porcelain"], {
      timeout: 5000,
    });
    if (!result.success) {
      return [];
    }

    const worktrees: Array<{ path: string; name: string; locked: boolean }> = [];
    let currentPath: string | null = null;
    let currentLocked = false;

    for (const line of result.stdout.split("\n")) {
      if (line.startsWith("worktree ")) {
        // Save previous worktree if exists
        if (currentPath !== null) {
          worktrees.push({
            path: currentPath,
            name: basename(currentPath),
            locked: currentLocked,
          });
        }
        currentPath = line.slice(9);
        currentLocked = false;
      } else if (line.startsWith("locked")) {
        currentLocked = true;
      }
    }

    // Don't forget the last worktree
    if (currentPath !== null) {
      worktrees.push({
        path: currentPath,
        name: basename(currentPath),
        locked: currentLocked,
      });
    }

    return worktrees;
  } catch {
    return [];
  }
}

/**
 * Execute git diff and return file list.
 * Returns empty array on error (fail silently).
 */
async function getGitDiffFiles(args: string[], cwd: string): Promise<string[]> {
  try {
    const result = await asyncSpawn("git", args, {
      timeout: 5000,
      cwd,
    });
    if (!result.success) {
      return [];
    }
    return result.stdout.trim().split("\n").filter(Boolean);
  } catch {
    return [];
  }
}

/**
 * Get changed files in a worktree compared to origin default branch.
 *
 * Combines two sources:
 * 1. `git diff --name-only origin/main...HEAD` - committed changes
 * 2. `git diff --name-only HEAD` - uncommitted changes (staged/unstaged)
 *
 * This ensures both committed and uncommitted changes are detected.
 */
export async function getWorktreeChangedFiles(worktreePath: string): Promise<string[]> {
  const originBranch = await getOriginDefaultBranch(worktreePath);
  const [committed, uncommitted] = await Promise.all([
    getGitDiffFiles(["diff", "--name-only", `${originBranch}...HEAD`], worktreePath),
    getGitDiffFiles(["diff", "--name-only", "HEAD"], worktreePath),
  ]);
  return [...new Set([...committed, ...uncommitted])];
}

/**
 * Extract issue number from worktree name.
 */
function extractIssueNumber(worktreeName: string): string | null {
  const match = worktreeName.match(/issue-(\d+)/i);
  return match ? match[1] : null;
}

/**
 * Get active worktree sessions (simplified version).
 * Uses lock status from getWorktreeList() to avoid N+1 queries.
 */
async function getActiveWorktreeSessions(
  _currentSessionId: string,
  _transcriptPath: string | null,
): Promise<{ sibling: WorktreeInfo[] }> {
  const siblings: WorktreeInfo[] = [];

  // Get current worktree path
  let currentWorktreePath: string | null = null;
  try {
    const result = await asyncSpawn("git", ["rev-parse", "--show-toplevel"], {
      timeout: 5000,
    });
    if (!result.success) {
      return { sibling: [] };
    }
    currentWorktreePath = result.stdout.trim();
  } catch {
    return { sibling: [] };
  }

  // Get all worktrees (includes lock status)
  const worktrees = await getWorktreeList();

  for (const wt of worktrees) {
    // Skip current worktree
    if (wt.path === currentWorktreePath) {
      continue;
    }

    // Skip main repo (not a worktree)
    if (!wt.path.includes(".worktrees")) {
      continue;
    }

    // Only include locked worktrees (active sessions)
    if (!wt.locked) {
      continue;
    }

    const changedFiles = await getWorktreeChangedFiles(wt.path);
    if (changedFiles.length === 0) {
      continue;
    }

    siblings.push({
      path: wt.path,
      issueNumber: extractIssueNumber(wt.name),
      changedFiles,
    });
  }

  return { sibling: siblings };
}

/**
 * Find worktrees that have the same file in their changed files.
 */
function findConflictingWorktrees(
  targetFile: string,
  activeSessions: { sibling: WorktreeInfo[] },
): ConflictInfo[] {
  const conflicts: ConflictInfo[] = [];
  const normalizedTarget = normalizePath(targetFile);

  // Check sibling worktrees for conflicts
  for (const info of activeSessions.sibling) {
    // Check if target file is in this worktree's changed files
    for (const changedFile of info.changedFiles) {
      if (normalizePath(changedFile) === normalizedTarget) {
        conflicts.push({
          issueNumber: info.issueNumber,
          path: info.path,
          changedFiles: info.changedFiles.sort().slice(0, 5),
          totalFiles: info.changedFiles.length,
        });
        break;
      }
    }
  }

  return conflicts;
}

/**
 * Format the conflict warning message.
 */
function formatWarning(targetFile: string, conflicts: ConflictInfo[]): string {
  const lines = ["⚠️ 並行編集の競合可能性:\n"];
  lines.push(`編集対象: ${normalizePath(targetFile)}\n`);
  lines.push("同一ファイルを編集中のsibling session:");

  for (const conflict of conflicts) {
    const issueStr = conflict.issueNumber
      ? `Issue #${conflict.issueNumber}`
      : (conflict.path.split("/").pop() ?? conflict.path);
    lines.push(`  - ${issueStr}`);
    lines.push(`    Worktree: ${conflict.path}`);

    const files = conflict.changedFiles;
    const totalFiles = conflict.totalFiles;
    let filesStr = files.slice(0, 3).join(", ");
    if (totalFiles > 3) {
      filesStr += ` (+${totalFiles - 3} more)`;
    }
    lines.push(`    変更中: ${filesStr}`);
    lines.push("");
  }

  lines.push("マージ時にコンフリクトが発生する可能性があります。");
  lines.push("Tip: 競合を避けるため、siblingセッションと調整するか、");
  lines.push("     別の独立したIssueに着手することを検討してください。");

  return lines.join("\n");
}

async function main(): Promise<void> {
  const result: { continue: boolean; systemMessage?: string } = { continue: true };

  try {
    const hookInput = await parseHookInput();
    const sessionId = hookInput.session_id ?? "";
    const source = hookInput.source ?? "";
    const transcriptPath = hookInput.transcript_path ?? null;
    const toolInput = (hookInput.tool_input ?? {}) as Record<string, unknown>;

    // Only run for fork-sessions
    if (!isForkSession(sessionId, source, transcriptPath)) {
      await logHookExecution(HOOK_NAME, "approve", "Not a fork-session");
      console.log(JSON.stringify(result));
      return;
    }

    // Get target file
    const targetFile = getTargetFile(toolInput);
    if (!targetFile) {
      await logHookExecution(HOOK_NAME, "approve", "No target file");
      console.log(JSON.stringify(result));
      return;
    }

    // Get active worktree sessions
    let activeSessions: { sibling: WorktreeInfo[] };
    try {
      activeSessions = await getActiveWorktreeSessions(sessionId, transcriptPath);
    } catch {
      // Fail silently - don't block on errors
      await logHookExecution(HOOK_NAME, "approve", "Error getting sessions");
      console.log(JSON.stringify(result));
      return;
    }

    // Find conflicts
    const conflicts = findConflictingWorktrees(targetFile, activeSessions);

    if (conflicts.length > 0) {
      const warning = formatWarning(targetFile, conflicts);
      result.systemMessage = warning;

      await logHookExecution(
        HOOK_NAME,
        "approve",
        `Warning: ${conflicts.length} conflicting worktree(s)`,
        { target_file: targetFile, conflicts },
      );
    } else {
      await logHookExecution(HOOK_NAME, "approve", "No conflicts");
    }
  } catch (error) {
    const errorMsg = `Hook error: ${formatError(error)}`;
    console.error(`[${HOOK_NAME}] ${errorMsg}`);
    await logHookExecution(HOOK_NAME, "approve", errorMsg);
  }

  console.log(JSON.stringify(result));
}

if (import.meta.main) {
  main();
}
